File: src/main/java/com/mparticle/kits/FlurryKit.java
package com.mparticle.kits;

import android.content.Context;
import android.content.Intent;
import android.location.Location;
import android.util.Log;

import com.flurry.android.Constants;
import com.flurry.android.FlurryAgent;
import com.flurry.android.FlurryInstallReceiver;
import com.mparticle.MPEvent;
import com.mparticle.MParticle;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class FlurryKit extends KitIntegration implements KitIntegration.AttributeListener, KitIntegration.EventListener {
    private static final String API_KEY = "apiKey";
    private static final String HASH_ID = "hashCustomerId";
    private static final String CAPTURE_EXCEPTIONS = "captureExceptions";
    private static final String INCLUDE_LOCATION = "includeLocation";
    private boolean includeLocation = true;

    @Override
    public String getName() {
        return "Flurry";
    }

    @Override
    protected List<ReportingMessage> onKitCreate(Map<String, String> settings, Context context) {
        boolean logEnabled = false;
        if (MParticle.getInstance().getEnvironment().equals(MParticle.Environment.Development)) {
            logEnabled = true;
        }
        if (getSettings().containsKey(INCLUDE_LOCATION) && !Boolean.parseBoolean(getSettings().get(INCLUDE_LOCATION))) {
            includeLocation = false;
        }

        new FlurryAgent.Builder()
                .withLogEnabled(logEnabled)
                .withLogLevel(Log.VERBOSE)
                .withCaptureUncaughtExceptions(Boolean.parseBoolean(getSettings().get(CAPTURE_EXCEPTIONS)))
                .build(context, getSettings().get(API_KEY));

        return null;
    }

    @Override
    public void setInstallReferrer(Intent intent) {
        new FlurryInstallReceiver().onReceive(getContext(), intent);
    }

    @Override
    public void setLocation(Location location) {
        if (includeLocation) {
            FlurryAgent.setLocation((float) location.getLatitude(), (float) location.getLongitude());
        }
    }

    @Override
    public void setUserIdentity(MParticle.IdentityType identityType, String id) {
       if (identityType.equals(MParticle.IdentityType.CustomerId)) {
           if (!getSettings().containsKey(HASH_ID) || Boolean.parseBoolean(getSettings().get(HASH_ID))){
               id = KitUtils.hashFnv1a(id.getBytes()).toString();
           }
           FlurryAgent.setUserId(id);
       }
    }

    @Override
    public void removeUserIdentity(MParticle.IdentityType identityType) {
        if (identityType.equals(MParticle.IdentityType.CustomerId)) {
            FlurryAgent.setUserId("");
        }
    }

    @Override
    public List<ReportingMessage> logout() {
        return null;
    }

    @Override
    public void setUserAttribute(String key, String value) {
        if (key.equals(MParticle.UserAttributes.AGE)) {
            try {
                FlurryAgent.setAge(Integer.parseInt(value));
            } catch (NumberFormatException nfe) {

            }
        }
        if(key.equals(MParticle.UserAttributes.GENDER)) {
            boolean female = value.toLowerCase().equalsIgnoreCase("female");
            FlurryAgent.setGender(female ? Constants.FEMALE : Constants.MALE);
        }
    }

    @Override
    public void setUserAttributeList(String s, List<String> list) {

    }

    @Override
    public boolean supportsAttributeLists() {
        return false;
    }

    @Override
    public void setAllUserAttributes(Map<String, String> attributes, Map<String, List<String>> attributeLists) {
        for (Map.Entry<String, String> entry : attributes.entrySet()){
            setUserAttribute(entry.getKey(), entry.getValue());
        }
    }

    @Override
    public void removeUserAttribute(String key) {
        //there's no way to un-set age/gender
    }

    @Override
    public List<ReportingMessage> leaveBreadcrumb(String breadcrumb) {
        return null;
    }

    @Override
    public List<ReportingMessage> logError(String message, Map<String, String> errorAttributes) {
        return null;
    }

    @Override
    public List<ReportingMessage> logException(Exception exception, Map<String, String> exceptionAttributes, String message) {
        return null;
    }

    @Override
    public List<ReportingMessage> logEvent(MPEvent event) {
        if (event.getInfo() == null) {
            FlurryAgent.logEvent(event.getEventName());
        }else{
            FlurryAgent.logEvent(event.getEventName(), event.getInfo());
        }
        List<ReportingMessage> messageList = new LinkedList<ReportingMessage>();
        messageList.add(ReportingMessage.fromEvent(this,event));
        return messageList;
    }

    @Override
    public List<ReportingMessage> logScreen(String screenName, Map<String, String> eventAttributes) {
        FlurryAgent.onPageView();
        List<ReportingMessage> messageList = new LinkedList<ReportingMessage>();
        messageList.add(
                new ReportingMessage(this, ReportingMessage.MessageType.SCREEN_VIEW, System.currentTimeMillis(), eventAttributes)
                        .setScreenName(screenName)
        );
        return messageList;
    }

    @Override
    public List<ReportingMessage> setOptOut(boolean optOutStatus) {
        FlurryAgent.setLogEvents(!optOutStatus);
        List<ReportingMessage> messageList = new LinkedList<ReportingMessage>();
        messageList.add(
                new ReportingMessage(this, ReportingMessage.MessageType.OPT_OUT, System.currentTimeMillis(), null)
                        .setOptOut(optOutStatus)
        );
        return messageList;
    }
}

File: src/main/kotlin/com/mparticle/kits/FlurryKit.kt
package com.mparticle.kits

import android.content.Context
import android.content.Intent
import android.util.Log
import com.flurry.android.Constants
import com.flurry.android.FlurryAgent
import com.flurry.android.FlurryInstallReceiver
import com.mparticle.MPEvent
import com.mparticle.MParticle
import com.mparticle.MParticle.IdentityType
import com.mparticle.kits.KitIntegration.AttributeListener
import java.util.*

class FlurryKit : KitIntegration(), AttributeListener, KitIntegration.EventListener {
    override fun getName(): String = KIT_NAME

    override fun onKitCreate(
        settings: Map<String, String>,
        context: Context
    ): List<ReportingMessage> {
        var logEnabled = false
        if (MParticle.getInstance()?.environment == MParticle.Environment.Development) {
            logEnabled = true
        }
        FlurryAgent.setReportLocation(
            getSettings().containsKey(INCLUDE_LOCATION) && !getSettings()[INCLUDE_LOCATION].toBoolean()
        )
        getSettings()[API_KEY]?.let {
            FlurryAgent.Builder()
                .withLogEnabled(logEnabled)
                .withLogLevel(Log.VERBOSE)
                .withCaptureUncaughtExceptions(getSettings()[CAPTURE_EXCEPTIONS].toBoolean())
                .build(context, it)
        }
        return emptyList()
    }

    override fun setInstallReferrer(intent: Intent) {
        FlurryInstallReceiver().onReceive(context, intent)
    }

    override fun setUserIdentity(identityType: IdentityType, idIn: String) {
        var id = idIn
        if (identityType == IdentityType.CustomerId) {
            if (!settings.containsKey(HASH_ID) || settings[HASH_ID].toBoolean()) {
                id = KitUtils.hashFnv1a(id.toByteArray()).toString()
            }
            FlurryAgent.setUserId(id)
        }
    }

    override fun removeUserIdentity(identityType: IdentityType) {
        if (identityType == IdentityType.CustomerId) {
            FlurryAgent.setUserId("")
        }
    }

    override fun logout(): List<ReportingMessage> = emptyList()

    override fun setUserAttribute(key: String, value: String) {
        if (key == MParticle.UserAttributes.AGE) {
            try {
                FlurryAgent.setAge(value.toInt())
            } catch (nfe: NumberFormatException) {
            }
        }
        if (key == MParticle.UserAttributes.GENDER) {
            val female = value.lowercase(Locale.getDefault()).equals("female", true)
            FlurryAgent.setGender(if (female) Constants.FEMALE else Constants.MALE)
        }
    }

    override fun setUserAttributeList(s: String, list: List<String>) {}
    override fun supportsAttributeLists(): Boolean = false

    override fun setAllUserAttributes(
        attributes: Map<String, String>,
        attributeLists: Map<String, List<String>>
    ) {
        for ((key, value) in attributes) {
            setUserAttribute(key, value)
        }
    }

    override fun removeUserAttribute(key: String) {
        //there's no way to un-set age/gender
    }

    override fun leaveBreadcrumb(breadcrumb: String): List<ReportingMessage> = emptyList()

    override fun logError(
        message: String,
        errorAttributes: Map<String, String>
    ): List<ReportingMessage> = emptyList()

    override fun logException(
        exception: Exception,
        exceptionAttributes: Map<String, String>,
        message: String
    ): List<ReportingMessage> = emptyList()

    override fun logEvent(event: MPEvent): List<ReportingMessage> {
        if (event.customAttributes == null) {
            FlurryAgent.logEvent(event.eventName)
        } else {
            event.customAttributeStrings?.let { FlurryAgent.logEvent(event.eventName, it) }
        }
        val messageList = LinkedList<ReportingMessage>()
        messageList.add(ReportingMessage.fromEvent(this, event))
        return messageList
    }

    override fun logScreen(
        screenName: String,
        eventAttributes: Map<String, String>
    ): List<ReportingMessage> {
        FlurryAgent.logEvent(screenName, eventAttributes)
        val messageList = LinkedList<ReportingMessage>()
        messageList.add(
            ReportingMessage(
                this,
                ReportingMessage.MessageType.SCREEN_VIEW,
                System.currentTimeMillis(),
                eventAttributes
            )
                .setScreenName(screenName)
        )
        return messageList
    }

    override fun setOptOut(optOutStatus: Boolean): List<ReportingMessage> = emptyList()

    companion object {
        private const val API_KEY = "apiKey"
        private const val HASH_ID = "hashCustomerId"
        private const val CAPTURE_EXCEPTIONS = "captureExceptions"
        private const val INCLUDE_LOCATION = "includeLocation"
        const val KIT_NAME = "Flurry"
    }
}
